### ISSUE TITLE
Can currently leak private types and values

### ISSUE THREAD

**Issue by baszalmstra** [mun-lang/mun#309]
Reexporting a type or value from one module to another should check the visibility of the original item.

This can happen in a few cases:

```rust
//- foo.mun
// Illegal, Bar has a smaller scope that this use statement
pub(super) struct Bar;

// Illegal, Bar has a smaller scope than this function
pub fn baz() -> Bar {
}
```

```rust
//- mod.mun
pub use foo::Bar; // Should be illegal because Bar is only visible to this module
```

In rust this emits the error: `can't leak private type`. This is currently not checked in Mun.

**Comment by mkhan45** [comment]
We also need to check that a public struct does not contain any private types in its exported fields:

```rs
struct Foo;

pub struct Bar {
    pub foo: Foo // illegal since Foo has a smaller scope than Bar
}
```

**Comment by mkhan45** [comment]
this is closed by #338 and #344

**Comment by baszalmstra** [comment]
Nice! Thanks @mkhan45 !
